### PR TITLE
deprecate unescaped shell special chars in commands

### PR DESCRIPTION
**Pull request by StefanKarpinski**

### CODE REVIEW

File: base/managers.jl [hunk 1]
```diff
@@ -179,10 +179,10 @@ function launch_on_machine(manager::SSHManager, machine, cnt, params, launched,
 
     # the default worker timeout
     tval = haskey(ENV, "JULIA_WORKER_TIMEOUT") ?
-        `export JULIA_WORKER_TIMEOUT=$(ENV["JULIA_WORKER_TIMEOUT"]);` : ``
+        `export JULIA_WORKER_TIMEOUT=$(ENV["JULIA_WORKER_TIMEOUT"])\;` : ``
 
     # Julia process with passed in command line flag arguments
-    cmd = `cd $dir && $tval $exename $exeflags`
+    cmd = `cd $dir '&&' $tval $exename $exeflags`
```
Comment (tkelman) [inline base/managers.jl]: this is in the ssh code, do we test this?

Comment (StefanKarpinski) [inline base/managers.jl]: I'm not sure. @amitmurthy, @tanmaykm?

Comment (tkelman) [inline base/managers.jl]: might be only behind the full-tests env var flag? or remote only...

we could add some simple tests that sending quoted/escaped special chars to programs that should be able to use them still act as expected

Comment (amitmurthy) [inline base/managers.jl]: If you wish to test, set `JULIA_TESTFULL=1` in your env. Enable password less ssh login into your local machine. Run the parallel test. This will test SSH Manager.

Else I can test this locally in a couple of days.

Comment (amitmurthy) [inline base/managers.jl]: `addprocs` hangs on the `open` call with this change. Printing `cmd` just before  the `open` shows that `&&` is not treated correctly.
```
`ssh -T -a -x -o ClearAllForwardings=yes -n localhost "sh -l -c \"cd /Users/amitm/Julia/julia '&&' /Users/amitm/Julia/julia/usr/bin/julia --worker ehqbRgqspml7nZLg\""`
```

**Comment by StefanKarpinski** [comment]
Ok, @amitmurthy – this seems to work for me locally although the output from `make test-parallel` with `export JULIA_TESTFULL=1` is a bit confusing. Let me know if it looks good to you.

File: base/managers.jl [hunk 2]
```diff
@@ -195,13 +195,14 @@ function launch_on_machine(manager::SSHManager, machine, cnt, params, launched,
...
     #      Used when running ssh in the background.
-    cmd = `ssh -T -a -x -o ClearAllForwardings=yes -n $sshflags $host $(shell_escape(cmd))`
+    cmd = `ssh -T -a -x -o ClearAllForwardings=yes -n $sshflags $host $(shell_escape(cmd, special = ""))`
 
     # launch the remote Julia process
 
     # detach launches the command in a new process group, allowing it to outlive
     # the initial julia process (Ctrl-C and teardown methods are handled through messages)
     # for the launched porcesses.
+    @show cmd
```
Comment (amitmurthy) [inline base/managers.jl]: Needs to be removed.

**Comment by amitmurthy** [comment]
Yeah, the output prints a few expected errors and one of the reasons it is not part of mainstream CI.

Test passes locally for me too.
 
Is the Travis error related to this?

A comment explaining the use of `shell_escape` would be a welcome addition.

File: base/managers.jl (hunk 2 above)
Comment (StefanKarpinski) [inline base/managers.jl]: oops, right :)

**Comment by StefanKarpinski** [comment]
The Travis failure is unrelated since I'm seeing it in multiple ongoing PRs. Something recently merged broke 32-bit Linux.

**Comment by tkelman** [comment]
agreed it's going to be a bit hard to follow when this needs to be bypassed

**Comment by StefanKarpinski** [comment]
It's not an exported function. I nevertheless added a docstring. If we have a way of representing pipelines like `test whatever && do something` then those could print correctly. Currently, we're abusing single-command syntax to pass constructs like that to shells. Hence the confusion.

**Comment by tkelman** [comment]
Should these be getting passed as strings instead then in this kind of situation?

**Comment by StefanKarpinski** [comment]
They could, but that doesn't really address the issue, which is that we don't currently have a way to represent command pipelines. Until we do, which requires a deprecation release cycle, anything we do is a hack. I could change the default for `shell_escape` to preserve the current behavior, but then command objects with special characters will print in a way that they're not supposed to be input. We could use a different default for commands and arrays of strings, but that just seems more confusing.

**Comment by StefanKarpinski** [comment]
Let me summarize:

1. Yes, this escaping situation is a bit confusing.
2. However, none of these functions is exported from Base.
3. The non-public behavior is documented in the docstring that's been added.
4. This situation is temporary: in 1.0 you will be able to write `test cond && do thing` and it will produce a pipeline object which will print the same way it's input and thus offer a correct and non-confusing way to express this (i.e. this code can go back the way it was).

**Comment by tkelman** [comment]
Sure. In this particular case though, the command in question is used as an argument to sh, so unless we implement a full sh parser and semantics, a Julia Cmd isn't the right type for holding argument data.

**Comment by StefanKarpinski** [comment]
We only need enough shell syntax to express the pipeline that's used here.

**Comment by StefanKarpinski** [comment]
Something got messed up here – either I forget to add some of the changes or I just messed this up. More changes are needed and are on their way here.

**Comment by tkelman** [comment]
This should leave a todo note in deprecated.jl.

**Comment by StefanKarpinski** [comment]
About what?

**Comment by tkelman** [comment]
That the deprecation should be removed so it doesn't get forgotten. This kind of came out of nowhere in #19598 without a lot of discussion about the future plan, so it should be tracked one way or another.

**Comment by StefanKarpinski** [comment]
The special characters are going to mean something in 1.0 – namely they're going to construct pipelines, etc. There has been discussion about it previously. No need for a note since this is going to have a new meaning.

**Comment by tkelman** [comment]
> There has been discussion about it previously

Not that I can recall, any more recently than `pipeline` was introduced and prior to #19598? There should be a 1.0-milestone issue for it then.

**Comment by StefanKarpinski** [comment]
I've definitely discussed with Jeff, Keno and Jameson but maybe it was not on GitHub. Issue opened: https://github.com/JuliaLang/julia/issues/20401.
